server/src/modules/jam.tsx

```diff
@@ -0,0 +1,234 @@
+/// <reference types="@kitajs/html/htmx.d.ts" />
+
+import Html from '@kitajs/html';
+
+import express from 'express';
+import {WebSocket} from 'ws';
+
+import {getChordsFromKeyRoot} from '../utils/music_utils';
+
+export const jamRouter = express.Router();
+
+jamRouter.use(express.json());
+jamRouter.use(express.urlencoded({extended: true}));
+
+enum ROUTES {
+    JAM_ACTIONS_ADD_CHORD = '/jam/actions/add_chord',
+    JAM_ACTIONS_NEW_JAM = '/jam/actions/new_jam',
+    JAM_ACTIONS_CHANGE_KEY = '/jam/actions/change_key',
+    JAM_ROOM_WEBSOCKET = '/jam/ws',
+}
+
+type JamState = {
+    key: string;
+    selectedChords: string[];
+};
+
+const jamState: JamState = {
+    key: 'C',
+    selectedChords: [],
+};
+
+const connectedSockets: WebSocket[] = [];
+
+export const initJamRouterWebsocket = () => {
+    jamRouter.ws(ROUTES.JAM_ROOM_WEBSOCKET, (ws, req) => {
+        connectedSockets.push(ws);
+        console.log('ws connected');
+        ws.send(JamView().toString());
+
+        ws.on('message', (msg) => {
+            console.log(msg);
+        });
+
+        ws.on('close', () => {
+            console.log('ws closed');
+            connectedSockets.splice(connectedSockets.indexOf(ws), 1);
+        });
+    });
+};
+
+const refreshAll = () => {
+    for (const ws of connectedSockets) {
+        ws.send(JamView().toString());
+    }
+};
+
+jamRouter.post<undefined, JSX.Element, undefined, {chord: string}>(
+    ROUTES.JAM_ACTIONS_ADD_CHORD,
+    (req, res) => {
+        const {chord} = req.query;
+        jamState.selectedChords.push(chord);
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+jamRouter.post<undefined, JSX.Element>(
+    ROUTES.JAM_ACTIONS_NEW_JAM,
+    (req, res) => {
+        jamState.selectedChords = [];
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+jamRouter.post<undefined, JSX.Element, {key: string}>(
+    ROUTES.JAM_ACTIONS_CHANGE_KEY,
+    (req, res) => {
+        const {key} = req.body;
+        jamState.key = key;
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+export const renderJamPage = () => {
+    return JamPage();
+};
+
+export const JamPage = () => {
+    return (
+        <>
+            <div hx-ws='connect:/jam/ws' />
```

Comment:
Also want to decouple the URL structure away from the views/components. There should be a functions like:

```ts
const makeJamViewWebsocketUrl = () => `connect:/jam/ws`;

const makeAddChordUrl = (chord: string) => `/jam/add_chord?chord=${chord}`;
```

I think all URL things should be done with functions like these, so there is no knowledge of the URL themselves in the view. Much easier to refactor.

These functions will be defined right next to the handlers that process the requests for that URL. They can use some other defined enum, but I prefer seeing the URL as a magic string when looking at the handler. No other code needs to know the URL structure, as that other code will just use the `makeAddChordUrl` function etc.